### PR TITLE
added gocd functions around pipeline reading and sending to GoCD

### PR DESCRIPTION
**Pull request by frank0631**

### CODE REVIEW

File: gocd/test.json [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      "secure": false,
+      "name": "PGSSLMODE",
+      "value": "require"
+    }
+  ],
+  "materials": [
+    {
+      "type": "git",
+      "attributes": {
+        "url": "https://github.cfpb.gov/databases/test.git",
```
Comment (marcesher) [inline gocd/test.json]: Probably shouldn't have internal URLs / IPs in a public repo

**Comment by marcesher** [comment]
Please spend some time on the README, so that it's clearer what this thing even is. I understand that it's just getting started; still, we should practice good hygiene

File: gocd/test.json (hunk 1 above)
Comment (tanderegg) [inline gocd/test.json]: +1 @frank0631 can you replace that with a public url?  Something like https://github.com/cfpb/rhobot-test.git?

**Comment by tanderegg** [comment]
Also +1, @frank0631 just provide a brief description of Rhobot, feel free to pull directly from the design document.

File: gocd/gocd.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	return UnmarshalPipeline(data)
+}
+
+func pipelineConfigPOST(gocdULR string, pipelineConfig PipelineConfig) Pipeline {
+
+	// Generated by curl-to-Go: https://mholt.github.io/curl-to-go
+
+	payloadBytes, err := json.Marshal(pipelineConfig)
+	if err != nil {
+		// handle err
```
Comment (tanderegg) [inline gocd/gocd.go]: @frank0631 I'd prefer not to commit code that does no error handling at all, could you look into providing some baseline handling of these cases?  It probably would make the most sense for this function to return a Pipeline and an Error, in which case you'd just return (nil, err), but you could also panic with the error message for now if you wish.  See: https://golang.org/doc/effective_go.html#errors

File: gocd/test.json (hunk 1 above)
Comment (frank0631) [inline gocd/test.json]: ip's and cfpb internal url's have been replaced

File: gocd/gocd.go (hunk 2 above)
Comment (frank0631) [inline gocd/gocd.go]: errors now bubble up from functions in the gocd pkg. An example of catching the errors can be see in the gocd_test.go

**Comment by frank0631** [comment]
The README and other non-code files that need to be updated will be in a seperate pull request from this pull request: https://github.com/cfpb/rhobot/pull/3

File: gocd/gocd.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	err = json.Indent(&prettyJSON, body, "", "\t")
+	if err != nil {
+		return pipeline, err
+	}
+	//fmt.Println("pipelineConfigGET:", string(prettyJSON.Bytes()))
+
+	return unmarshalPipeline(body)
+}
+
+func Push(gocdULR string, path string, group string) {
```
Comment (tanderegg) [inline gocd/gocd.go]: Go likes to have all public functions contain a line of comment, starting with the function name, like // Push ....
We should follow that convention.

File: gocd/gocd.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+func ReadPipelineJSONFromFile(path string) (pipeline Pipeline, err error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return pipeline, err
+	}
+	return unmarshalPipeline(data)
+}
+
+func pipelineConfigPOST(gocdULR string, pipelineConfig PipelineConfig) (pipeline Pipeline, err error) {
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL not gocdULR

File: gocd/gocd.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+	// Generated by curl-to-Go: https://mholt.github.io/curl-to-go
+
+	payloadBytes, err := json.Marshal(pipelineConfig)
+	if err != nil {
+		return pipeline, err
+	}
+	payloadBody := bytes.NewReader(payloadBytes)
+
+	req, err := http.NewRequest("POST", gocdULR+"/go/api/admin/pipelines", payloadBody)
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return pipeline, err
+	}
+
+	var prettyJSON bytes.Buffer
+	err = json.Indent(&prettyJSON, body, "", "\t")
+	if err != nil {
+		return pipeline, err
+	}
+	//fmt.Println("pipelineConfigPOST:", string(prettyJSON.Bytes()))
```
Comment (tanderegg) [inline gocd/gocd.go:141]: Preferable to using comments, we should be able to run in a formal "debug" or "verbose" mode.  Look into Logrus or  glog?

File: gocd/gocd.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+	return unmarshalPipeline(body)
+}
+
+func pipelineGET(gocdULR string, pipelineName string) (pipeline Pipeline, err error) {
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+	// Generated by curl-to-Go: https://mholt.github.io/curl-to-go
+
+	req, err := http.NewRequest("GET", gocdULR+"/go/api/admin/pipelines/"+pipelineName, nil)
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return pipeline, err
+	}
+
+	var prettyJSON bytes.Buffer
+	err = json.Indent(&prettyJSON, body, "", "\t")
+	if err != nil {
+		return pipeline, err
+	}
+	//fmt.Println("pipelineConfigGET:", string(prettyJSON.Bytes()))
```
Comment (tanderegg) [inline gocd/gocd.go:176]: See above comment by logging

File: gocd/gocd.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+	return unmarshalPipeline(body)
+}
+
+func Push(gocdULR string, path string, group string) {
+
+	pipeline, err := ReadPipelineJSONFromFile(path)
+	check(err)
+	pipelineConfig := PipelineConfig{group, pipeline}
+	_, err = pipelineConfigPOST(gocdULR, pipelineConfig)
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+	check(err)
+}
+
+func Pull(gocdULR string, path string) {
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+	pipeline, err := ReadPipelineJSONFromFile(path)
+	check(err)
+	name := pipeline.Name
+
+	Clone(gocdULR , path , name )
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func Clone(gocdULR string, path string, name string) {
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd.go [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+	pipelineFetched, err := pipelineGET(gocdULR, name)
```
Comment (tanderegg) [inline gocd/gocd.go]: gocdURL

File: gocd/gocd_test.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
+package gocd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	//"github.com/davecgh/go-spew/spew"
```
Comment (tanderegg) [inline gocd/gocd_test.go]: Let's remove commented spew code for now

File: gocd/test.json [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+            "secure": false,
+            "name": "PGSSLMODE",
+            "value": "require"
+        }
+    ],
+    "materials": [
+        {
+            "type": "git",
+            "attributes": {
+                "url": "https://github.com/cfpb/rhobot-test.git?",
```
Comment (tanderegg) [inline gocd/test.json]: Remove unneeded "?"

Comment (frank0631) [inline gocd/test.json]: the "?" was in the test pipeline in GoCD. I've modified the pipeline and pulled it

**Comment by tanderegg** [comment]
Looks good, thanks @frank0631!

**Pull request merged**